src/passes/Precompute.cpp

```diff
@@ -282,18 +283,16 @@ struct Precompute
     localGraph.computeInfluences();
     // prepare the work list. we add things here that might change to a constant
     // initially, that means everything
-    std::unordered_set<Expression*> work;
+    UniqueDeferredQueue<Expression*> work;
     for (auto& pair : localGraph.locations) {
       auto* curr = pair.first;
-      work.insert(curr);
+      work.push(curr);
     }
     // the constant value, or none if not a constant
     std::unordered_map<LocalSet*, Literals> setValues;
     // propagate constant values
     while (!work.empty()) {
-      auto iter = work.begin();
-      auto* curr = *iter;
-      work.erase(iter);
+      auto* curr = work.pop();
```

Comment:
Is this the same semantics as the `unordered_set::erase`? It looks if `UniqueDeferredQueue` contains a count for each item and the item is removed only if the count reaches zero, whereas in `unordered_set` the item is erased right away. 

At the same time I don't think I have a full grasp on when `UniqueDeferredQueue` should be used...

Reply:
The semantics should be the same, yes. Erase removes the item from a set, and a `UniqueDeferredQueue` has a queue of unique items - so it is effectively like a set in that popping an item removes it and there cannot be others like it.

In general `UniqueDeferredQueue` is nice because it has a simpler API (replacing the 3 lines here with one), and also it is more "deterministic" in that using an `unordered_set` will cause the order to change based on the user's c++ standard hashing. A `UniqueDeferredQueue` operates as a queue, so the order is always the same (with clear semantics for how to handle a push of an existing item - it is deferred to the end; it could be either way).

Reply:
> The semantics should be the same, yes. Erase removes the item from a set, and a `UniqueDeferredQueue` has a queue of unique items - so it is effectively like a set in that popping an item removes it and there cannot be others like it.

So it is guaranteed there are no duplicates pushed to the queue in this code? If there are duplicates the semantics can be different right?

Reply:
It handles duplicates itself. If the someone pushes `A, B, C, D, B` then `B` will still only be popped once. In this example the first pop will be `A`, then `C` (*not* `B`, which is deferred), then `D`, and only then `B`, and then the queue is empty.

Reply:
Ah I see. I was misunderstanding how `UniqueDeferredQueue` works.